### PR TITLE
Convert SHFileOperation error to Win IO error

### LINKED ISSUE #45398: wrong error messages in DeleteFile
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/wrong-error-messages-in-DeleteFile/1146912)._

---
<p>Function Microsoft.VisualBasic.FileIO.FileSystem.DeleteFile has wrong error messages if it is used to delete file to the recycle bin. If any IOException occurs, then exception message is nonsense. For example it throws exception &quot;This function is not supported on this system.&quot; if current user does not have permissions. If a network path is invalid, it throws exception &quot;The system call level is not correct.&quot;. I think the problem is that it internally calls Win32 API function SHFileOperation which has some special error numbers.</p>

---
#### Feedback Bot on 8/12/2020, 07:25 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 11/20/2020, 01:42 AM: 

I detected that this issue hasn’t received a lot of activity, votes, or comments in the past 90 days. Based on this, the issues severity and affected area, it’s my experience that this issue is unlikely to get fixed. To improve the situation, consider following best practices for quality problem reports, and giving us more details on how this issue is impacting you.

#### Dada Huo [MSFT] on 11/25/2020, 00:01 AM: 

<p>Thank you for taking the time to log this issue! Did you still reproduce this issue in the latest stable version of Visual Studio installed from <a target='_blank' href="https://visualstudio.microsoft.com/downloads">https://visualstudio.microsoft.com/downloads</a> ?<br>
If so, could you please provide a sample solution that has this problem so that we can conduct further research? It would be very helpful if you can provide a screenshot of this issue.</p>
<p>We look forward to hearing from you!</p>

#### Petr Laštovička on 11/30/2020, 03:52 AM: 

<p>I have installed the latest Visual Studio and .Net 5.<br>
This is a sample C# 9 project:</p>
<pre><code><span class="hljs-keyword">using</span> Microsoft.VisualBasic.FileIO;
<span class="hljs-keyword">using</span> System;
<span class="hljs-keyword">using</span> System.Windows.Forms;

<span class="hljs-keyword">try</span>
{
  FileSystem.DeleteFile(@<span class="hljs-string">"C:\Windows\System32\kernel32.dll"</span>, UIOption.OnlyErrorDialogs, RecycleOption.SendToRecycleBin);
}
<span class="hljs-keyword">catch</span> (Exception ex)
{
  MessageBox.Show(ex.Message);
}
</code></pre>
<p>I would expect error that access is denied. But it shows message “This function is not supported on this system.”</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by ehasis**
Convert SHFileOperation error code to Win IO error code.

Fix  #45398

### CODE REVIEW

**Comment by ehasis** [comment]
Not sure how to write a proper test for this code. I managed to test on my machine just like pointed in the original issue, I also tried setting the access control, but the API for this is only available when adding System.IO.FileSystem.AccessControl. The result also differs from .NET Framework.

**Comment by danmoseley** [comment]
A manual test is probably OK. This code is rarely changed.
Are these the only codes it makes sense to map? We know that many others will result in bogus behavior. Perhaps those are unlikely cases.

**Comment by ehasis** [comment]
Those make sense for FileDelete. I will try find out and manual test for others that is related do Copy/Move.

**Comment by danmoseley** [comment]
As we know for sure any that we don't translate are just wrong, I think it would also be fine you just go through the list in the docs and try to map anything reasonable even if you can't produce it locally.

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb [hunk 1]
```diff
@@ -1905,6 +1905,70 @@ Namespace Microsoft.VisualBasic.FileIO
...
+                Case NativeTypes.DE_FILENAMETOOLONG
+                    Return NativeTypes.ERROR_BUFFER_OVERFLOW
+                Case NativeTypes.DE_DEST_IS_CDROM
+                    Return NativeTypes.ERROR_WRITE_FAULT
+                Case NativeTypes.DE_DEST_IS_DVD
+                    Return NativeTypes.ERROR_WRITE_FAULT
+                Case NativeTypes.DE_DEST_IS_CDRECORD
+                    Return NativeTypes.ERROR_WRITE_FAULT
+                Case NativeTypes.DE_FILE_TOO_LARGE
+                    Return NativeTypes.ERROR_DISK_FULL
```
Comment (cston) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb:1951]: >ERROR_DISK_FULL [](http://example.com/codeflow?start=39&length=15)

Should we use `ERROR_FILE_TOO_LARGE = 223` here?

**Review by cston — APPROVED** [review]

File: src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb [hunk 2]
```diff
@@ -1905,6 +1905,70 @@ Namespace Microsoft.VisualBasic.FileIO
...
+                Case NativeTypes.DE_ROOTDIR
+                    Return NativeTypes.ERROR_ACCESS_DENIED
+                Case NativeTypes.DE_OPCANCELLED
+                    Return NativeTypes.ERROR_CANCELLED
+                Case NativeTypes.DE_DESTSUBTREE
+                    Return NativeTypes.ERROR_BAD_PATHNAME
+                Case NativeTypes.DE_ACCESSDENIEDSRC
+                    Return NativeTypes.ERROR_ACCESS_DENIED
+                Case NativeTypes.DE_PATHTOODEEP
+                    Return NativeTypes.ERROR_BUFFER_OVERFLOW
```
Comment (danmoseley) [inline src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb:1931]: ERROR_FILENAME_EXCED_RANGE ?

**Review by cston — APPROVED** [review]

**Comment by danmoseley** [comment]
Thanks

**Pull request merged**